sw/device/silicon_creator/lib/drivers/kmac.h

```diff
@@ -0,0 +1,126 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_DRIVERS_KMAC_H_
+#define OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_DRIVERS_KMAC_H_
+
+/**
+ * This driver is specialized to meet the needs of SPHINCS+-SHAKE.
+ */
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Configure the KMAC block at startup.
+ *
+ * Sets the KMAC block to use software entropy (since we have no secret inputs
+ * for SPHINCS+) and sets the mode to SHAKE-256.
+ *
+ * @return Error code indicating if the operation succeeded.
+ */
+OT_WARN_UNUSED_RESULT
+rom_error_t kmac_shake256_configure(void);
+
+/**
+ * Start a SHAKE-256 hashing operation.
+ *
+ * Must be called after `kmac_shake256_configure()`. Will block until KMAC
+ * hardware is idle.
+ *
+ * This driver supports SHAKE-256 hashing with the following pattern:
+ * - Exactly one call to `kmac_shake256_start`
+ * - Zero or more calls to `kmac_shake256_absorb`
+ * - Exactly one call to `kmac_shake256_squeeze_start`
+ * - Exactly one call to `kmac_shake256_squeeze_end`
```

Comment:
do you think it would be useful to define a wrapper that does all of this? This, of course, only makes sense if we will only hash contiguous regions.

`kmac_shake256(const void *buf, size_t buf_len, size_t digest_word_cnt, uint32_t *digest)`

Reply:
All of the hashing in SPHINCS+ combines multiple non-contiguous regions, so it's helpful to have the separate `absorb` calls. The reference implementation does use a single `shake256()` call in some places, but it has to copy the input to contiguous memory first, which seems wasteful and hurts performance.

Reply:
Thanks for the explanation, for a second I thought this would be used like generic hash function over a contiguous buffer.